compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/ParameterizeGlobals.cpp

```diff
@@ -0,0 +1,248 @@
+#include "iree/hal/api.h"
+#include "iree/io/formats/irpa/irpa_builder.h"
+#include "iree/tooling/parameter_util.h"
+
+#include "iree/compiler/Dialect/Stream/IR/StreamDialect.h"
+#include "iree/compiler/Dialect/Stream/IR/StreamTypes.h"
+#include "iree/compiler/Dialect/Util/IR/UtilDialect.h"
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "iree/compiler/Modules/IO/Parameters/Transforms/PassDetail.h"
+#include "iree/compiler/Modules/IO/Parameters/Transforms/Passes.h"
+#include "llvm/ADT/ScopeExit.h"
+
+namespace mlir::iree_compiler::IREE::IO::Parameters {
+
+namespace {
+
+static void iree_io_file_handle_release_mapping(
+    void *user_data, iree_io_file_handle_primitive_t handle_primitive) {
+  iree_file_contents_free((iree_file_contents_t *)user_data);
+}
+
+static iree_status_t iree_tooling_open_output_parameter_file(
+    iree_io_physical_offset_t archive_offset,
+    iree_io_physical_size_t archive_length, iree_allocator_t host_allocator,
+    iree_io_file_handle_t **out_file_handle, const char *path) {
+  iree_file_contents_t *file_contents = NULL;
+  IREE_RETURN_IF_ERROR(iree_file_create_mapped(
+      path, archive_offset + archive_length, archive_offset,
+      (iree_host_size_t)archive_length, host_allocator, &file_contents));
+  iree_io_file_handle_release_callback_t release_callback = {
+      iree_io_file_handle_release_mapping, file_contents};
+  iree_status_t status = iree_io_file_handle_wrap_host_allocation(
+      IREE_IO_FILE_ACCESS_WRITE, file_contents->buffer, release_callback,
+      host_allocator, out_file_handle);
+  if (!iree_status_is_ok(status)) {
+    iree_file_contents_free(file_contents);
+  }
+  return status;
+}
+
+struct ParameterizeGlobalsPass
+    : public ParameterizeGlobalsBase<ParameterizeGlobalsPass> {
+  ParameterizeGlobalsPass(std::string archivePathStr,
+                          std::string parameterNamespaceStr) {
+    this->archivePath = archivePathStr;
+    this->parameterNamespace = parameterNamespaceStr;
+  };
+  ParameterizeGlobalsPass(const ParameterizeGlobalsPass &pass)
+      : ParameterizeGlobalsPass(pass.archivePath, pass.parameterNamespace) {}
+
+  void runOnOperation() override {
+    // Nothing to do if no path specified.
+    if (archivePath.empty()) {
+      return;
+    }
+
+    MLIRContext *context = &getContext();
+    ModuleOp moduleOp = getOperation();
+
+    iree_allocator_t host_allocator = iree_allocator_system();
+
+    // Create the parameter archive builder.
+    iree_io_parameter_archive_builder_t builder;
+    iree_io_parameter_archive_builder_initialize(host_allocator, &builder);
+
+    auto deinitializeExit = llvm::make_scope_exit([&]() {
+      return iree_io_parameter_archive_builder_deinitialize(&builder);
+    });
+
+    SmallVector<IREE::Util::GlobalOp> constantGlobals;
+    // Walk the globals in the module.
+    for (auto global : moduleOp.getOps<IREE::Util::GlobalOp>()) {
+      // Mutable globals cannot become parameters.
```

Comment:
they can both for initialization purposes (read initial value from parameters then mutate in memory) and for mutable parameters (not fully implemented yet). worth calling out here that this is a TODO and something we'll fix.